api/src/main/java/org/openmrs/module/sync/SyncUtil.java

```diff
@@ -289,212 +302,202 @@ public static Object newObject(String className) throws Exception {
-				if (log.isDebugEnabled())
+				if (log.isDebugEnabled()) {
 					log.debug("Now inspecting superclass: " + superClazz.getName());
-				
-				superFields = superClazz.getDeclaredFields();
-				if (superFields != null) {
-					for (Field f : superFields) {
-						allFields.add(f);
-					}
+				}
+				for (Field f : superClazz.getDeclaredFields()) {
+					allFields.put(f.getName(), f);
 				}
 				superClazz = superClazz.getSuperclass();
 			}
-			if (nativeFields != null) {
-				// add native fields
-				for (Field f : nativeFields) {
-					allFields.add(f);
-				}
+			// add native fields
+			for (Field f :  clazz.getDeclaredFields()) {
+				allFields.put(f.getName(), f);
 			}
 		}
-		
 		return allFields;
 	}
 	
 	public static OpenmrsObject getOpenmrsObj(String className, String uuid) {
 		try {
-			OpenmrsObject o = Context.getService(SyncService.class).getOpenmrsObjectByUuid(
-			    (Class<OpenmrsObject>) Context.loadClass(className), uuid);
-			
-			if (log.isDebugEnabled()) {
-				if (o == null) {
-					log.debug("Unable to get an object of type " + className + " with Uuid " + uuid + ";");
-				}
-			}
-			return o;
+			Class<? extends OpenmrsObject> clazz = (Class<? extends OpenmrsObject>) Context.loadClass(className);
+			return getOpenmrsObj(clazz, uuid);
 		}
 		catch (ClassNotFoundException ex) {
 			log.warn("getOpenmrsObj couldn't find class: " + className, ex);
 			return null;
 		}
 	}
+
+	public static OpenmrsObject getOpenmrsObj(Class<? extends OpenmrsObject> clazz, String uuid) {
+		OpenmrsObject o = Context.getService(SyncService.class).getOpenmrsObjectByUuid(clazz, uuid);
+		if (log.isDebugEnabled()) {
+			if (o == null) {
+				log.debug("Unable to get an object of type " + clazz.getName() + " with Uuid " + uuid + ";");
+			}
+		}
+		return o;
+	}
 	
-	public static Object valForField(String fieldName, String fieldVal, ArrayList<Field> allFields, Node n) {
+	public static Object valForField(Field field, String xmlFieldValue, String xmlFieldType) {
+
+		// If the given fieldType is a primitive or a String, handle these directly
+		switch (xmlFieldType) {
+			case "int": { return Integer.valueOf(xmlFieldValue); }
+			case "long": { return Long.valueOf(xmlFieldValue); }
+			case "double": { return Double.valueOf(xmlFieldValue); }
+			case "float": { return Float.valueOf(xmlFieldValue); }
+			case "boolean": { return Boolean.valueOf(xmlFieldValue); }
+			case "byte": { return Byte.valueOf(xmlFieldValue); }
+			case "short": { return Short.valueOf(xmlFieldValue); }
+		}
+
+		// Determine the type of field this is
+		Class<?> classType = null;
+		if (field != null) {
+			classType = getFieldType(field);
+			if (classType == null) {
+				log.debug("type for " + field.getName() + " unable to be determined from field reflection");
+			}
+		}
+		else {
+			log.debug("field is null, trying to get class from xmlFieldType: " + xmlFieldType);
+		}
+		if (classType == null) {

```

Comment:
This block here is the main addition that got things working with PatientProgramAttributes.